File: generated_code/cpp/solution_671_3.cpp
#include <algorithm>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 * int val;
 * TreeNode *left;
 * TreeNode *right;
 * TreeNode() : val(0), left(nullptr), right(nullptr) {}
 * TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    /**
     * Algorithm Choice: Optimized DFS
     * * Key Property: The root value is the absolute minimum of the entire tree.
     * To find the second minimum, we look for the smallest value strictly greater than root->val.
     * * Pruning: If a node's value is already greater than the root's value, that node (and its entire subtree) 
     * cannot contain a smaller "second minimum candidate" than itself, because the root of any subtree
     * is the minimum of that subtree. Thus, we don't need to traverse deeper once we find a value > root->val.
     * * Time Complexity: O(N) in worst case (traversing all nodes), though often less due to pruning.
     * Space Complexity: O(H) where H is the height of the tree (recursion stack).
     */
    int findSecondMinimumValue(TreeNode* root) {
        if (!root) return -1;
        return dfs(root, root->val);
    }

private:
    long dfs(TreeNode* node, int minVal) {
        // Base case: if we reach a null or find a value greater than the tree minimum
        if (!node) return -1;
        
        if (node->val > minVal) {
            // This node is a candidate for the second minimum.
            // Since subtrees only contain values >= this node's value, we stop here.
            return node->val;
        }
        
        // If node->val == minVal, we must check its children to find a larger value
        long left = dfs(node->left, minVal);
        long right = dfs(node->right, minVal);
        
        // If both children return valid second minimum candidates, take the smaller one
        if (left != -1 && right != -1) {
            return std::min(left, right);
        }
        
        // If only one child (or neither) returned a valid candidate, return the non- -1 value
        return (left != -1) ? left : right;
    }
};

File: generated_code/python/solution_671_3.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def findSecondMinimumValue(self, root: Optional[TreeNode]) -> int:
        """
        Time Complexity: O(N) in the worst case, where N is the number of nodes.
                         We prune subtrees where the root value is already greater 
                         than the current global minimum, avoiding unnecessary traversals.
        Space Complexity: O(H) where H is the height of the tree, representing the recursion stack.
        """
        if not root:
            return -1
        
        # The root of the tree always holds the absolute minimum value of the entire tree.
        min_val = root.val
        self.second_min = float('inf')
        
        def dfs(node: Optional[TreeNode]) -> None:
            if not node:
                return
            
            # If the current node's value is greater than the absolute minimum,
            # it is a candidate for the second minimum value.
            if min_val < node.val < self.second_min:
                self.second_min = node.val
                
            # Crucial Pruning: If a node's value is greater than or equal to the 
            # currently found second minimum, its children cannot yield a smaller 
            # second minimum due to the tree's heap-like property (root = min(left, right)).
            if node.val < self.second_min:
                dfs(node.left)
                dfs(node.right)
                
        dfs(root)
        
        # If second_min was never updated, it means all nodes have the same value.
        return self.second_min if self.second_min != float('inf') else -1